File: src/lib/integrations/google-calendar.ts
import crypto from "node:crypto";
import fs from "node:fs/promises";
import path from "node:path";

type OAuthStateRecord = {
  state: string;
  createdAt: string;
  returnTo: string | null;
};

type TokenStore = {
  accessToken: string;
  refreshToken: string | null;
  scope: string | null;
  tokenType: string | null;
  expiresAt: string | null;
  updatedAt: string;
};

export type GoogleCalendarEvent = {
  id: string;
  summary: string;
  status: string;
  htmlLink: string | null;
  location: string | null;
  start: string | null;
  end: string | null;
};

const DEFAULT_SHARED_ROOT = process.env.SHARED_ROOT_PATH?.trim() || path.join(process.cwd(), "shared_data");
const OAUTH_STATE_TTL_MS = 10 * 60 * 1000;

function parseBoolean(raw: string | undefined, fallback: boolean): boolean {
  if (!raw) {
    return fallback;
  }
  const token = raw.trim().toLowerCase();
  if (token === "1" || token === "true" || token === "yes" || token === "on") {
    return true;
  }
  if (token === "0" || token === "false" || token === "no" || token === "off") {
    return false;
  }
  return fallback;
}

function resolveStorePath(rawPath: string | undefined, fileName: string): string {
  const fallback = path.join(DEFAULT_SHARED_ROOT, "shared_memory", fileName);
  const configured = rawPath?.trim();
  if (!configured) {
    return fallback;
  }
  if (path.isAbsolute(configured)) {
    return configured;
  }
  return path.join(process.cwd(), configured);
}

const TOKEN_PATH = resolveStorePath(process.env.GOOGLE_CALENDAR_TOKEN_PATH, "google_calendar_tokens.json");
const STATE_PATH = resolveStorePath(process.env.GOOGLE_CALENDAR_STATE_PATH, "google_calendar_oauth_state.json");

async function ensureParentDir(filePath: string) {
  await fs.mkdir(path.dirname(filePath), { recursive: true });
}

async function readJsonFile<T>(filePath: string, fallback: T): Promise<T> {
  try {
    const raw = await fs.readFile(filePath, "utf-8");
    return JSON.parse(raw) as T;
  } catch {
    return fallback;
  }
}

async function writeJsonFile(filePath: string, payload: unknown) {
  await ensureParentDir(filePath);
  const tempPath = `${filePath}.tmp`;
  await fs.writeFile(tempPath, JSON.stringify(payload, null, 2), "utf-8");
  await fs.rename(tempPath, filePath);
}

function requiredEnv(name: string): string {
  const value = process.env[name]?.trim();
  if (!value) {
    throw new Error(`missing_env:${name}`);
  }
  return value;
}

function getOAuthConfig() {
  return {
    clientId: requiredEnv("GOOGLE_CALENDAR_OAUTH_CLIENT_ID"),
    clientSecret: requiredEnv("GOOGLE_CALENDAR_OAUTH_CLIENT_SECRET"),
    redirectUri: requiredEnv("GOOGLE_CALENDAR_OAUTH_REDIRECT_URI"),
    scope:
      process.env.GOOGLE_CALENDAR_OAUTH_SCOPES?.trim() || "https://www.googleapis.com/auth/calendar.readonly",
  };
}

export function isGoogleCalendarEnabled() {
  return parseBoolean(process.env.GOOGLE_CALENDAR_ENABLED, false);
}

export function isGoogleCalendarReadonly() {
  return parseBoolean(process.env.GOOGLE_CALENDAR_READONLY, true);
}

export async function createGoogleOAuthState(returnTo: string | null): Promise<OAuthStateRecord> {
  const record: OAuthStateRecord = {
    state: crypto.randomUUID(),
    createdAt: new Date().toISOString(),
    returnTo: returnTo?.trim() ? returnTo.trim() : null,
  };
  await writeJsonFile(STATE_PATH, record);
  return record;
}

export async function consumeGoogleOAuthState(state: string): Promise<OAuthStateRecord | null> {
  const stored = await readJsonFile<OAuthStateRecord | null>(STATE_PATH, null);
  if (!stored || stored.state !== state) {
    return null;
  }
  const createdAt = Date.parse(stored.createdAt);
  if (!Number.isFinite(createdAt)) {
    return null;
  }
  if (Date.now() - createdAt > OAUTH_STATE_TTL_MS) {
    return null;
  }
  await fs.rm(STATE_PATH, { force: true });
  return stored;
}

export function buildGoogleOAuthAuthorizationUrl(state: string): string {
  const config = getOAuthConfig();
  const url = new URL("https://accounts.google.com/o/oauth2/v2/auth");
  url.searchParams.set("client_id", config.clientId);
  url.searchParams.set("redirect_uri", config.redirectUri);
  url.searchParams.set("response_type", "code");
  url.searchParams.set("scope", config.scope);
  url.searchParams.set("access_type", "offline");
  url.searchParams.set("include_granted_scopes", "true");
  url.searchParams.set("prompt", "consent");
  url.searchParams.set("state", state);
  return url.toString();
}

async function exchangeAuthorizationCode(code: string): Promise<TokenStore> {
  const config = getOAuthConfig();
  const body = new URLSearchParams({
    client_id: config.clientId,
    client_secret: config.clientSecret,
    code,
    grant_type: "authorization_code",
    redirect_uri: config.redirectUri,
  });
  const response = await fetch("https://oauth2.googleapis.com/token", {
    method: "POST",
    headers: { "content-type": "application/x-www-form-urlencoded" },
    body: body.toString(),
  });
  if (!response.ok) {
    const detail = await response.text();
    throw new Error(`google_token_exchange_failed:${response.status}:${detail}`);
  }
  const data = (await response.json()) as {
    access_token?: string;
    refresh_token?: string;
    scope?: string;
    token_type?: string;
    expires_in?: number;
  };
  const accessToken = (data.access_token ?? "").trim();
  if (!accessToken) {
    throw new Error("google_token_exchange_missing_access_token");
  }
  const expiresAt =
    typeof data.expires_in === "number" && Number.isFinite(data.expires_in)
      ? new Date(Date.now() + data.expires_in * 1000).toISOString()
      : null;
  return {
    accessToken,
    refreshToken: data.refresh_token?.trim() || null,
    scope: data.scope?.trim() || null,
    tokenType: data.token_type?.trim() || "Bearer",
    expiresAt,
    updatedAt: new Date().toISOString(),
  };
}

async function refreshAccessToken(existing: TokenStore): Promise<TokenStore> {
  if (!existing.refreshToken) {
    throw new Error("google_refresh_token_missing");
  }
  const config = getOAuthConfig();
  const body = new URLSearchParams({
    client_id: config.clientId,
    client_secret: config.clientSecret,
    refresh_token: existing.refreshToken,
    grant_type: "refresh_token",
  });
  const response = await fetch("https://oauth2.googleapis.com/token", {
    method: "POST",
    headers: { "content-type": "application/x-www-form-urlencoded" },
    body: body.toString(),
  });
  if (!response.ok) {
    const detail = await response.text();
    throw new Error(`google_token_refresh_failed:${response.status}:${detail}`);
  }
  const data = (await response.json()) as {
    access_token?: string;
    scope?: string;
    token_type?: string;
    expires_in?: number;
  };
  const accessToken = (data.access_token ?? "").trim();
  if (!accessToken) {
    throw new Error("google_token_refresh_missing_access_token");
  }
  const expiresAt =
    typeof data.expires_in === "number" && Number.isFinite(data.expires_in)
      ? new Date(Date.now() + data.expires_in * 1000).toISOString()
      : null;
  const refreshed: TokenStore = {
    accessToken,
    refreshToken: existing.refreshToken,
    scope: data.scope?.trim() || existing.scope,
    tokenType: data.token_type?.trim() || existing.tokenType || "Bearer",
    expiresAt,
    updatedAt: new Date().toISOString(),
  };
  await writeJsonFile(TOKEN_PATH, refreshed);
  return refreshed;
}

function tokenExpired(tokens: TokenStore): boolean {
  if (!tokens.expiresAt) {
    return false;
  }
  const expiresAt = Date.parse(tokens.expiresAt);
  if (!Number.isFinite(expiresAt)) {
    return false;
  }
  return Date.now() >= expiresAt - 60_000;
}

async function getStoredTokens(): Promise<TokenStore | null> {
  return readJsonFile<TokenStore | null>(TOKEN_PATH, null);
}

export async function saveGoogleTokenFromCode(code: string): Promise<TokenStore> {
  const token = await exchangeAuthorizationCode(code);
  await writeJsonFile(TOKEN_PATH, token);
  return token;
}

async function getValidAccessToken(): Promise<string> {
  const stored = await getStoredTokens();
  if (!stored || !stored.accessToken) {
    throw new Error("google_calendar_not_connected");
  }
  if (!tokenExpired(stored)) {
    return stored.accessToken;
  }
  const refreshed = await refreshAccessToken(stored);
  return refreshed.accessToken;
}

function normalizeIsoDate(value: string | null): string | null {
  if (!value) {
    return null;
  }
  const parsed = Date.parse(value);
  if (!Number.isFinite(parsed)) {
    return null;
  }
  return new Date(parsed).toISOString();
}

function buildTodayWindow() {
  const now = new Date();
  const start = new Date(now);
  start.setHours(0, 0, 0, 0);
  const end = new Date(now);
  end.setHours(23, 59, 59, 999);
  return {
    timeMin: start.toISOString(),
    timeMax: end.toISOString(),
  };
}

async function fetchCalendarEvents(accessToken: string, calendarId: string, timeMin: string, timeMax: string) {
  const url = new URL(
    `https://www.googleapis.com/calendar/v3/calendars/${encodeURIComponent(calendarId)}/events`
  );
  url.searchParams.set("singleEvents", "true");
  url.searchParams.set("orderBy", "startTime");
  url.searchParams.set("timeMin", timeMin);
  url.searchParams.set("timeMax", timeMax);
  url.searchParams.set("maxResults", "30");
  const response = await fetch(url, {
    headers: { Authorization: `Bearer ${accessToken}` },
  });
  if (!response.ok) {
    const detail = await response.text();
    throw new Error(`google_calendar_events_failed:${response.status}:${detail}`);
  }
  return (await response.json()) as { items?: Array<Record<string, unknown>> };
}

export async function listGoogleTodayEvents(input: {
  calendarId?: string | null;
  timeMin?: string | null;
  timeMax?: string | null;
}): Promise<{
  calendarId: string;
  timeMin: string;
  timeMax: string;
  events: GoogleCalendarEvent[];
}> {
  const calendarId = input.calendarId?.trim() || process.env.GOOGLE_CALENDAR_ID?.trim() || "primary";
  const window = buildTodayWindow();
  const timeMin = normalizeIsoDate(input.timeMin ?? null) || window.timeMin;
  const timeMax = normalizeIsoDate(input.timeMax ?? null) || window.timeMax;

  let accessToken = await getValidAccessToken();
  let raw: { items?: Array<Record<string, unknown>> };
  try {
    raw = await fetchCalendarEvents(accessToken, calendarId, timeMin, timeMax);
  } catch (error) {
    if (!(error instanceof Error) || !error.message.startsWith("google_calendar_events_failed:401:")) {
      throw error;
    }
    const stored = await getStoredTokens();
    if (!stored) {
      throw error;
    }
    accessToken = (await refreshAccessToken(stored)).accessToken;
    raw = await fetchCalendarEvents(accessToken, calendarId, timeMin, timeMax);
  }

  const events: GoogleCalendarEvent[] = (raw.items ?? []).map((item) => {
    const startValue = item.start as { dateTime?: string; date?: string } | undefined;
    const endValue = item.end as { dateTime?: string; date?: string } | undefined;
    return {
      id: String(item.id ?? ""),
      summary: String(item.summary ?? "(제목 없음)"),
      status: String(item.status ?? "unknown"),
      htmlLink: item.htmlLink ? String(item.htmlLink) : null,
      location: item.location ? String(item.location) : null,
      start: startValue?.dateTime ? String(startValue.dateTime) : startValue?.date ? String(startValue.date) : null,
      end: endValue?.dateTime ? String(endValue.dateTime) : endValue?.date ? String(endValue.date) : null,
    };
  });

  return { calendarId, timeMin, timeMax, events };
}

export async function getGoogleCalendarConnectionStatus() {
  const token = await getStoredTokens();
  return {
    enabled: isGoogleCalendarEnabled(),
    readonly: isGoogleCalendarReadonly(),
    connected: Boolean(token?.accessToken),
    tokenUpdatedAt: token?.updatedAt ?? null,
    tokenExpiresAt: token?.expiresAt ?? null,
    scope: token?.scope ?? null,
  };
}

File: proxy/app/google_calendar.py
from __future__ import annotations

import json
import os
import uuid
from datetime import datetime, timezone
from pathlib import Path
from urllib.parse import urlencode
from urllib.request import Request, urlopen
from zoneinfo import ZoneInfo


OAUTH_STATE_TTL_MS = 10 * 60 * 1000


def _parse_bool(raw: str | None, fallback: bool) -> bool:
    if not raw:
        return fallback
    token = raw.strip().lower()
    if token in {"1", "true", "yes", "on"}:
        return True
    if token in {"0", "false", "no", "off"}:
        return False
    return fallback


def _default_shared_root() -> Path:
    configured = (os.getenv("SHARED_ROOT_PATH") or "").strip()
    if configured:
        return Path(configured)
    return Path("/app/shared_data")


def _resolve_store_path(raw_path: str | None, file_name: str) -> Path:
    fallback = _default_shared_root() / "shared_memory" / file_name
    configured = (raw_path or "").strip()
    if not configured:
        return fallback
    path = Path(configured)
    if path.is_absolute():
        return path
    return Path.cwd() / configured


TOKEN_PATH = _resolve_store_path(os.getenv("GOOGLE_CALENDAR_TOKEN_PATH"), "google_calendar_tokens.json")
STATE_PATH = _resolve_store_path(os.getenv("GOOGLE_CALENDAR_STATE_PATH"), "google_calendar_oauth_state.json")


def _ensure_parent(path: Path) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)


def _read_json(path: Path, fallback):
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except Exception:  # noqa: BLE001
        return fallback


def _write_json(path: Path, payload) -> None:
    _ensure_parent(path)
    temp = path.with_suffix(path.suffix + ".tmp")
    temp.write_text(json.dumps(payload, ensure_ascii=False, indent=2), encoding="utf-8")
    try:
        temp.chmod(0o600)
    except Exception:  # noqa: BLE001
        pass
    temp.replace(path)
    try:
        path.chmod(0o600)
    except Exception:  # noqa: BLE001
        pass


def _required_env(name: str) -> str:
    value = (os.getenv(name) or "").strip()
    if not value:
        raise RuntimeError(f"missing_env:{name}")
    return value


def _oauth_config() -> dict[str, str]:
    return {
        "clientId": _required_env("GOOGLE_CALENDAR_OAUTH_CLIENT_ID"),
        "clientSecret": _required_env("GOOGLE_CALENDAR_OAUTH_CLIENT_SECRET"),
        "redirectUri": _required_env("GOOGLE_CALENDAR_OAUTH_REDIRECT_URI"),
        "scope": (os.getenv("GOOGLE_CALENDAR_OAUTH_SCOPES") or "").strip()
        or "https://www.googleapis.com/auth/calendar.readonly",
    }


def is_google_calendar_enabled() -> bool:
    return _parse_bool(os.getenv("GOOGLE_CALENDAR_ENABLED"), False)


def is_google_calendar_readonly() -> bool:
    return _parse_bool(os.getenv("GOOGLE_CALENDAR_READONLY"), True)


def create_google_oauth_state(return_to: str | None) -> dict[str, str | None]:
    record = {
        "state": str(uuid.uuid4()),
        "createdAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
        "returnTo": (return_to or "").strip() or None,
    }
    _write_json(STATE_PATH, record)
    return record


def consume_google_oauth_state(state: str) -> dict[str, str | None] | None:
    stored = _read_json(STATE_PATH, None)
    if not isinstance(stored, dict):
        return None
    if str(stored.get("state", "")) != state:
        return None

    created_at = str(stored.get("createdAt", ""))
    try:
        created_ms = int(datetime.fromisoformat(created_at.replace("Z", "+00:00")).timestamp() * 1000)
    except ValueError:
        return None
    now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
    if now_ms - created_ms > OAUTH_STATE_TTL_MS:
        return None

    try:
        STATE_PATH.unlink(missing_ok=True)
    except Exception:  # noqa: BLE001
        pass
    return {
        "state": str(stored.get("state", "")),
        "createdAt": created_at,
        "returnTo": stored.get("returnTo") if isinstance(stored.get("returnTo"), str) else None,
    }


def build_google_oauth_authorization_url(state: str) -> str:
    config = _oauth_config()
    query = urlencode(
        {
            "client_id": config["clientId"],
            "redirect_uri": config["redirectUri"],
            "response_type": "code",
            "scope": config["scope"],
            "access_type": "offline",
            "include_granted_scopes": "true",
            "prompt": "consent",
            "state": state,
        }
    )
    return f"https://accounts.google.com/o/oauth2/v2/auth?{query}"


def _post_form(url: str, payload: dict[str, str]) -> dict:
    body = urlencode(payload).encode("utf-8")
    request = Request(
        url,
        data=body,
        headers={"content-type": "application/x-www-form-urlencoded"},
        method="POST",
    )
    with urlopen(request, timeout=12) as response:
        raw = response.read().decode("utf-8")
    return json.loads(raw)


def _exchange_authorization_code(code: str) -> dict:
    config = _oauth_config()
    data = _post_form(
        "https://oauth2.googleapis.com/token",
        {
            "client_id": config["clientId"],
            "client_secret": config["clientSecret"],
            "code": code,
            "grant_type": "authorization_code",
            "redirect_uri": config["redirectUri"],
        },
    )
    access_token = str(data.get("access_token") or "").strip()
    if not access_token:
        raise RuntimeError("google_token_exchange_missing_access_token")
    expires_in = data.get("expires_in")
    expires_at = None
    if isinstance(expires_in, (int, float)):
        expires_at = datetime.fromtimestamp(
            datetime.now(timezone.utc).timestamp() + float(expires_in),
            tz=timezone.utc,
        ).isoformat().replace("+00:00", "Z")
    return {
        "accessToken": access_token,
        "refreshToken": str(data.get("refresh_token") or "").strip() or None,
        "scope": str(data.get("scope") or "").strip() or None,
        "tokenType": str(data.get("token_type") or "").strip() or "Bearer",
        "expiresAt": expires_at,
        "updatedAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
    }


def _token_expired(tokens: dict) -> bool:
    expires_at = tokens.get("expiresAt")
    if not isinstance(expires_at, str) or not expires_at:
        return False
    try:
        expires_ms = int(datetime.fromisoformat(expires_at.replace("Z", "+00:00")).timestamp() * 1000)
    except ValueError:
        return False
    now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
    return now_ms >= expires_ms - 60_000


def _stored_tokens() -> dict | None:
    payload = _read_json(TOKEN_PATH, None)
    return payload if isinstance(payload, dict) else None


def save_google_token_from_code(code: str) -> dict:
    token = _exchange_authorization_code(code)
    _write_json(TOKEN_PATH, token)
    return token


def _refresh_access_token(existing: dict) -> dict:
    refresh_token = str(existing.get("refreshToken") or "").strip()
    if not refresh_token:
        raise RuntimeError("google_refresh_token_missing")

    config = _oauth_config()
    data = _post_form(
        "https://oauth2.googleapis.com/token",
        {
            "client_id": config["clientId"],
            "client_secret": config["clientSecret"],
            "refresh_token": refresh_token,
            "grant_type": "refresh_token",
        },
    )
    access_token = str(data.get("access_token") or "").strip()
    if not access_token:
        raise RuntimeError("google_token_refresh_missing_access_token")

    expires_in = data.get("expires_in")
    expires_at = None
    if isinstance(expires_in, (int, float)):
        expires_at = datetime.fromtimestamp(
            datetime.now(timezone.utc).timestamp() + float(expires_in),
            tz=timezone.utc,
        ).isoformat().replace("+00:00", "Z")

    refreshed = {
        "accessToken": access_token,
        "refreshToken": refresh_token,
        "scope": str(data.get("scope") or existing.get("scope") or "").strip() or None,
        "tokenType": str(data.get("token_type") or existing.get("tokenType") or "").strip() or "Bearer",
        "expiresAt": expires_at,
        "updatedAt": datetime.now(timezone.utc).isoformat().replace("+00:00", "Z"),
    }
    _write_json(TOKEN_PATH, refreshed)
    return refreshed


def _valid_access_token() -> str:
    stored = _stored_tokens()
    if not stored:
        raise RuntimeError("google_calendar_not_connected")
    access_token = str(stored.get("accessToken") or "").strip()
    if not access_token:
        raise RuntimeError("google_calendar_not_connected")
    if not _token_expired(stored):
        return access_token
    refreshed = _refresh_access_token(stored)
    return str(refreshed.get("accessToken") or "")


def _normalize_iso_date(value: str | None) -> str | None:
    if not value:
        return None
    try:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    except ValueError:
        return None
    return parsed.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")


def _today_window() -> tuple[str, str]:
    configured_timezone = (os.getenv("GOOGLE_CALENDAR_TIMEZONE") or os.getenv("MINERVA_BRIEFING_TIMEZONE") or "Asia/Seoul").strip()
    try:
        local_tz = ZoneInfo(configured_timezone or "Asia/Seoul")
    except Exception:  # noqa: BLE001
        local_tz = ZoneInfo("Asia/Seoul")
    now = datetime.now(local_tz)
    start = now.replace(hour=0, minute=0, second=0, microsecond=0)
    end = now.replace(hour=23, minute=59, second=59, microsecond=999000)
    return (
        start.astimezone(timezone.utc).isoformat().replace("+00:00", "Z"),
        end.astimezone(timezone.utc).isoformat().replace("+00:00", "Z"),
    )


def _fetch_calendar_events(access_token: str, calendar_id: str, time_min: str, time_max: str) -> dict:
    query = urlencode(
        {
            "singleEvents": "true",
            "orderBy": "startTime",
            "timeMin": time_min,
            "timeMax": time_max,
            "maxResults": "30",
        }
    )
    url = f"https://www.googleapis.com/calendar/v3/calendars/{calendar_id}/events?{query}"
    request = Request(url, headers={"Authorization": f"Bearer {access_token}"}, method="GET")
    with urlopen(request, timeout=12) as response:
        raw = response.read().decode("utf-8")
    return json.loads(raw)


def list_google_today_events(
    input_payload: dict[str, str | None] | None = None,
) -> dict:
    input_payload = input_payload or {}
    calendar_id = (input_payload.get("calendarId") or os.getenv("GOOGLE_CALENDAR_ID") or "primary").strip()
    default_min, default_max = _today_window()
    time_min = _normalize_iso_date(input_payload.get("timeMin")) or default_min
    time_max = _normalize_iso_date(input_payload.get("timeMax")) or default_max

    access_token = _valid_access_token()
    try:
        raw = _fetch_calendar_events(access_token, calendar_id, time_min, time_max)
    except Exception as error:  # noqa: BLE001
        message = str(error)
        if "401" not in message:
            raise
        stored = _stored_tokens()
        if not stored:
            raise
        refreshed = _refresh_access_token(stored)
        raw = _fetch_calendar_events(str(refreshed.get("accessToken") or ""), calendar_id, time_min, time_max)

    events = []
    for item in raw.get("items", []) if isinstance(raw, dict) else []:
        if not isinstance(item, dict):
            continue
        start = item.get("start")
        end = item.get("end")
        start_value = start if isinstance(start, dict) else {}
        end_value = end if isinstance(end, dict) else {}
        events.append(
            {
                "id": str(item.get("id") or ""),
                "summary": str(item.get("summary") or "(제목 없음)"),
                "status": str(item.get("status") or "unknown"),
                "htmlLink": str(item.get("htmlLink")) if item.get("htmlLink") else None,
                "location": str(item.get("location")) if item.get("location") else None,
                "start": str(start_value.get("dateTime") or start_value.get("date") or "") or None,
                "end": str(end_value.get("dateTime") or end_value.get("date") or "") or None,
            }
        )

    return {
        "calendarId": calendar_id,
        "timeMin": time_min,
        "timeMax": time_max,
        "events": events,
    }


def get_google_calendar_connection_status() -> dict:
    token = _stored_tokens()
    token_expired = _token_expired(token) if isinstance(token, dict) else False
    refresh_available = bool(isinstance(token, dict) and str(token.get("refreshToken") or "").strip())
    return {
        "enabled": is_google_calendar_enabled(),
        "readonly": is_google_calendar_readonly(),
        "connected": bool(token and token.get("accessToken")),
        "tokenExpired": token_expired,
        "refreshAvailable": refresh_available,
        "tokenUpdatedAt": token.get("updatedAt") if isinstance(token, dict) else None,
        "tokenExpiresAt": token.get("expiresAt") if isinstance(token, dict) else None,
        "scope": token.get("scope") if isinstance(token, dict) else None,
    }
